File: src/render/meshing/FaceCulling.js
/**
 * Face culling utilities for voxel meshing
 * @module render/meshing/FaceCulling
 */

import { isLeafBlock } from '../../config/BlockConfig.js';
import { AIR, WATER, UNLOADED_BLOCK } from '../../core/constants.js';
import { buildBlockLookups } from '../../optimization/BlockLookups.js';

/**
 * Pre-built block lookup tables
 * @type {Object}
 */
let lookups = null;

/**
 * Get or build block lookups (lazy initialization)
 * @returns {Object} Block lookup tables
 */
function getLookups() {
    if (!lookups) {
        lookups = buildBlockLookups();
    }
    return lookups;
}

/**
 * Face direction offsets [dx, dy, dz, nx, ny, nz]
 * @type {Array<Array<number>>}
 */
export const FACE_DIRECTIONS = [
    [0, 1, 0, 0, 1, 0],   // Top (+Y)
    [0, -1, 0, 0, -1, 0], // Bottom (-Y)
    [1, 0, 0, 1, 0, 0],   // East (+X)
    [-1, 0, 0, -1, 0, 0], // West (-X)
    [0, 0, 1, 0, 0, 1],   // South (+Z)
    [0, 0, -1, 0, 0, -1], // North (-Z)
];

/**
 * Face index constants
 * @type {Object}
 */
export const FACE = {
    TOP: 0,
    BOTTOM: 1,
    EAST: 2,
    WEST: 3,
    SOUTH: 4,
    NORTH: 5,
};

/**
 * Check if a face should be rendered based on neighboring block
 * @param {number} blockId - Current block ID
 * @param {number} neighborId - Neighboring block ID
 * @returns {boolean} True if face should be rendered
 */
export function shouldRenderFace(blockId, neighborId) {
    // Don't render if current block is air
    if (blockId === AIR) return false;

    // If neighbor is unloaded, treat as solid (don't render)
    if (neighborId === UNLOADED_BLOCK) return false;

    const { BLOCK_IS_SOLID, IS_TRANSPARENT } = getLookups();

    // Don't render if neighbor is opaque solid
    if (BLOCK_IS_SOLID[neighborId] && !IS_TRANSPARENT[neighborId]) {
        return false;
    }

    // Water only draws faces against air or leaves
    if (blockId === WATER) {
        return neighborId === AIR || isLeafBlock(neighborId);
    }

    // For transparent blocks, check neighbor transparency
    return IS_TRANSPARENT[neighborId] === 1;
}

/**
 * Check if block is transparent (needs special handling)
 * @param {number} blockId - Block ID to check
 * @returns {boolean} True if transparent
 */
export function isTransparent(blockId) {
    return getLookups().IS_TRANSPARENT[blockId] === 1;
}

/**
 * Check if block is solid
 * @param {number} blockId - Block ID to check
 * @returns {boolean} True if solid
 */
export function isSolid(blockId) {
    return getLookups().BLOCK_IS_SOLID[blockId] === 1;
}

/**
 * Check if block is opaque (solid and not transparent)
 * @param {number} blockId - Block ID to check
 * @returns {boolean} True if opaque
 */
export function isOpaque(blockId) {
    const { BLOCK_IS_SOLID, IS_TRANSPARENT } = getLookups();
    return BLOCK_IS_SOLID[blockId] === 1 && IS_TRANSPARENT[blockId] === 0;
}

/**
 * Get neighbor offset for a face index
 * @param {number} faceIndex - Face index (0-5)
 * @returns {{dx: number, dy: number, dz: number, nx: number, ny: number, nz: number}}
 */
export function getFaceOffset(faceIndex) {
    const dir = FACE_DIRECTIONS[faceIndex];
    return {
        dx: dir[0],
        dy: dir[1],
        dz: dir[2],
        nx: dir[3],
        ny: dir[4],
        nz: dir[5],
    };
}

/**
 * Get all visible faces for a block
 * @param {number} blockId - Block ID
 * @param {Function} getNeighbor - Function to get neighbor block ID (dx, dy, dz) => blockId
 * @returns {Array<number>} Array of visible face indices
 */
export function getVisibleFaces(blockId, getNeighbor) {
    if (blockId === AIR) return [];

    const visibleFaces = [];
    for (let faceIdx = 0; faceIdx < 6; faceIdx++) {
        const { dx, dy, dz } = getFaceOffset(faceIdx);
        const neighborId = getNeighbor(dx, dy, dz);
        if (shouldRenderFace(blockId, neighborId)) {
            visibleFaces.push(faceIdx);
        }
    }
    return visibleFaces;
}

/**
 * Count visible faces for a block (optimized version without array allocation)
 * @param {number} blockId - Block ID
 * @param {Function} getNeighbor - Function to get neighbor block ID
 * @returns {number} Number of visible faces
 */
export function countVisibleFaces(blockId, getNeighbor) {
    if (blockId === AIR) return 0;

    let count = 0;
    for (let faceIdx = 0; faceIdx < 6; faceIdx++) {
        const { dx, dy, dz } = getFaceOffset(faceIdx);
        const neighborId = getNeighbor(dx, dy, dz);
        if (shouldRenderFace(blockId, neighborId)) {
            count++;
        }
    }
    return count;
}

export default {
    shouldRenderFace,
    isTransparent,
    isSolid,
    isOpaque,
    getFaceOffset,
    getVisibleFaces,
    countVisibleFaces,
    FACE_DIRECTIONS,
    FACE,
};

File: voxel_engine/engine/meshing/face_culling.py
"""
Face culling - determine which voxel faces are visible.

A face is visible (should be rendered) if the adjacent block is:
- Air or transparent
- Or current block is transparent and adjacent is a different type
"""

from typing import Generator, Tuple

from voxel_engine.engine.meshing.constants import (
    CHUNK_SIZE_X, CHUNK_SIZE_Y, CHUNK_SIZE_Z, FACE_DIRECTIONS, AIR_BLOCK
)
from voxel_engine.engine.state import WorldState
from voxel_engine.engine.registry import Registry


def should_render_face(
    world: WorldState,
    bx: int, by: int, bz: int,
    face_index: int,
    block_id: int
) -> bool:
    """
    Check if a face should be rendered.

    A face is visible if:
    - The adjacent block is air/transparent
    - OR the current block is transparent and adjacent is different

    Args:
        world: WorldState containing chunk data.
        bx: Global X coordinate of block.
        by: Global Y coordinate of block (0-319).
        bz: Global Z coordinate of block.
        face_index: Which face to check (0-5).
        block_id: ID of the block being checked.

    Returns:
        bool: True if face should be rendered.
    """
    # Get neighbor position
    dx = int(FACE_DIRECTIONS[face_index, 0])
    dy = int(FACE_DIRECTIONS[face_index, 1])
    dz = int(FACE_DIRECTIONS[face_index, 2])
    nx, ny, nz = bx + dx, by + dy, bz + dz

    # Out of world bounds - render face (sky/void)
    if ny < 0 or ny >= CHUNK_SIZE_Y:
        return True

    neighbor_id = world.get_block(nx, ny, nz)

    # Unloaded chunk returns 0 (AIR), which is fine
    # Air always shows face
    if neighbor_id == AIR_BLOCK:
        return True

    # Check transparency
    block_transparent = Registry.is_transparent(block_id)
    neighbor_transparent = Registry.is_transparent(neighbor_id)

    # Solid blocks: only render if neighbor is transparent
    if not block_transparent:
        return neighbor_transparent

    # Transparent blocks: render if neighbor is different type
    # (prevents z-fighting between same transparent blocks)
    return neighbor_id != block_id


def iter_visible_faces(
    world: WorldState,
    cx: int, cz: int
) -> Generator[Tuple[int, int, int, int, int], None, None]:
    """
    Iterate over all visible faces in a chunk.

    Yields tuples of (local_x, local_y, local_z, face_index, block_id)
    for each face that should be rendered.

    Args:
        world: WorldState containing chunk data.
        cx: Chunk X coordinate.
        cz: Chunk Z coordinate.

    Yields:
        Tuple[int, int, int, int, int]: (lx, ly, lz, face_index, block_id)
    """
    chunk = world.get_chunk(cx, cz)
    if chunk is None:
        return

    base_x = cx * CHUNK_SIZE_X
    base_z = cz * CHUNK_SIZE_Z

    # Iterate through all blocks in the chunk
    for ly in range(CHUNK_SIZE_Y):
        for lz in range(CHUNK_SIZE_Z):
            for lx in range(CHUNK_SIZE_X):
                # Get block at this position
                block_id = chunk.get_block(lx, ly, lz)

                # Skip air
                if block_id == AIR_BLOCK:
                    continue

                # Global coordinates for neighbor checks
                gx = base_x + lx
                gz = base_z + lz

                # Check each face
                for face_index in range(6):
                    if should_render_face(world, gx, ly, gz,
                                          face_index, block_id):
                        yield (lx, ly, lz, face_index, block_id)


def iter_visible_faces_fast(
    world: WorldState,
    cx: int, cz: int
) -> Generator[Tuple[int, int, int, int, int], None, None]:
    """
    Optimized version of iter_visible_faces using direct array access.

    Uses numpy array indexing for faster iteration through chunk data.

    Args:
        world: WorldState containing chunk data.
        cx: Chunk X coordinate.
        cz: Chunk Z coordinate.

    Yields:
        Tuple[int, int, int, int, int]: (lx, ly, lz, face_index, block_id)
    """
    chunk = world.get_chunk(cx, cz)
    if chunk is None:
        return

    blocks = chunk.blocks
    base_x = cx * CHUNK_SIZE_X
    base_z = cz * CHUNK_SIZE_Z

    # Cache registry lookups for performance
    is_transparent = Registry.is_transparent
    get_block = world.get_block

    # Iterate through all blocks (Y outer for cache locality)
    for ly in range(CHUNK_SIZE_Y):
        for lz in range(CHUNK_SIZE_Z):
            for lx in range(CHUNK_SIZE_X):
                # Direct array access is faster than method call
                block_id = int(blocks[lx, ly, lz])

                # Skip air
                if block_id == AIR_BLOCK:
                    continue

                # Global coordinates
                gx = base_x + lx
                gz = base_z + lz

                block_transparent = is_transparent(block_id)

                # Check each of 6 faces
                # Face 0: +X (East)
                nx = gx + 1
                neighbor_id = get_block(nx, ly, gz)
                if neighbor_id == AIR_BLOCK or (
                    is_transparent(neighbor_id) if not block_transparent
                    else neighbor_id != block_id
                ):
                    yield (lx, ly, lz, 0, block_id)

                # Face 1: -X (West)
                nx = gx - 1
                neighbor_id = get_block(nx, ly, gz)
                if neighbor_id == AIR_BLOCK or (
                    is_transparent(neighbor_id) if not block_transparent
                    else neighbor_id != block_id
                ):
                    yield (lx, ly, lz, 1, block_id)

                # Face 2: +Y (Up)
                ny = ly + 1
                if ny >= CHUNK_SIZE_Y:
                    yield (lx, ly, lz, 2, block_id)
                else:
                    neighbor_id = get_block(gx, ny, gz)
                    if neighbor_id == AIR_BLOCK or (
                        is_transparent(neighbor_id) if not block_transparent
                        else neighbor_id != block_id
                    ):
                        yield (lx, ly, lz, 2, block_id)

                # Face 3: -Y (Down)
                ny = ly - 1
                if ny < 0:
                    yield (lx, ly, lz, 3, block_id)
                else:
                    neighbor_id = get_block(gx, ny, gz)
                    if neighbor_id == AIR_BLOCK or (
                        is_transparent(neighbor_id) if not block_transparent
                        else neighbor_id != block_id
                    ):
                        yield (lx, ly, lz, 3, block_id)

                # Face 4: +Z (South)
                nz = gz + 1
                neighbor_id = get_block(gx, ly, nz)
                if neighbor_id == AIR_BLOCK or (
                    is_transparent(neighbor_id) if not block_transparent
                    else neighbor_id != block_id
                ):
                    yield (lx, ly, lz, 4, block_id)

                # Face 5: -Z (North)
                nz = gz - 1
                neighbor_id = get_block(gx, ly, nz)
                if neighbor_id == AIR_BLOCK or (
                    is_transparent(neighbor_id) if not block_transparent
                    else neighbor_id != block_id
                ):
                    yield (lx, ly, lz, 5, block_id)
